bleak/backends/bluezdbus/client.py

```diff
@@ -768,140 +915,93 @@ async def stop_notify(
             )
         )
 
-        if message.body[0] == defs.GATT_CHARACTERISTIC_INTERFACE:
-            if message.path in self._notification_callbacks:
-                logger.debug(
-                    "GATT Char Properties Changed: {0} | {1}".format(
-                        message.path, message.body[1:]
-                    )
+        if message.member == "InterfacesAdded":
+            path, interfaces = message.body
+
+            if defs.GATT_SERVICE_INTERFACE in interfaces:
+                obj = unpack_variants(interfaces[defs.GATT_SERVICE_INTERFACE])
+                # if this assert fails, it means our match rules are probably wrong
+                assert obj["Device"] == self._device_path
+                self.services.add_service(BleakGATTServiceBlueZDBus(obj, path))
+
+            if defs.GATT_CHARACTERISTIC_INTERFACE in interfaces:
+                obj = unpack_variants(interfaces[defs.GATT_CHARACTERISTIC_INTERFACE])
+                service = next(
+                    x
+                    for x in self.services.services.values()
+                    if x.path == obj["Service"]
                 )
-                self._notification_callbacks[message.path](
-                    message.path, message.body[1]
+                self.services.add_characteristic(
+                    BleakGATTCharacteristicBlueZDBus(obj, path, service.uuid)
                 )
-        elif message.body[0] == defs.DEVICE_INTERFACE:
-            device_path = "/org/bluez/%s/dev_%s" % (
-                self._adapter,
-                self.address.replace(":", "_"),
-            )
-            if message.path.lower() == device_path.lower():
-                message_body_map = message.body[1]
-                if (
-                    "Connected" in message_body_map
-                    and not message_body_map["Connected"]
-                ):
-                    logger.debug("Device {} disconnected.".format(self.address))
+
+            if defs.GATT_DESCRIPTOR_INTERFACE in interfaces:
+                obj = unpack_variants(interfaces[defs.GATT_DESCRIPTOR_INTERFACE])
+                handle = int(obj["Characteristic"][-4:], 16)
+                characteristic = self.services.characteristics[handle]
+                self.services.add_descriptor(
+                    BleakGATTDescriptorBlueZDBus(obj, path, characteristic.uuid, handle)
+                )
+        elif message.member == "InterfacesRemoved":
+            path, interfaces = message.body
+
+        elif message.member == "PropertiesChanged":
+            interface, changed, _ = message.body
+            changed = unpack_variants(changed)
+
+            if interface == defs.GATT_CHARACTERISTIC_INTERFACE:
+                if message.path in self._notification_callbacks and "Value" in changed:
+                    handle = int(message.path[-4:], 16)
+                    self._notification_callbacks[message.path](handle, changed["Value"])
```

Comment:
There are probably lots of places where I have done something like this (in other backends too) since I didn't realize that all of the APIs are supposed to be `bytearray`.

Reply:
I fixed this particular instance.